### DISCUSSION TITLE
OData Dynamic Properties query

### DISCUSSION THREAD

**Discussion by acidindustries** [Azure/azure-mobile-apps#748] [Q&A]
Looking for some guidance on this. 

I have the following class defined to store my items.

```csharp
public class Book : ITableData {
        public string Id {get; set;}
        public IDictionary<string, object> Metadata {get; set;}
}
```

My TableController is of type Book and returns the records with the metadata. OData specs say the I should be able to query on subdocuments (ex: $filter=Metadata/ReviewCount gt 1) but the query is failing. Is this supported right now? 

Basically trying to replicate what is going on [here](https://learn.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-v4/use-open-types-in-odata-v4) but querying the Metadata.

**Comment by adrianhall** [comment]
Azure Mobile Apps is not a generic OData controller - we use a restricted set of capabilities that supports the LINQ provider for the client. 
 You can't do that sort of query with our client, hence we don't support it in the server either.

Having said that, if we can get a suitable test framework for this, I think we could work on something.  Something like

``` csharp
var items = await table.Where(x => x.MetaData["ReviewCount"] > 1).ToListAsync();
```

However, it's not generalizable in LINQ, so it would be hard to do.  Open to thoughts here.

**Reply by acidindustries** [reply]
I had to come to the realization that since the controller isn't based on ODataController that it wouldn't work for the scenario. Is there an approach you have in mind for what you suggested? I might be able to implement it as a rough draft.

**Reply by amertlich** [reply]
I really like this idea! Maybe I'm reading into this to much, but as I understand the implication, it would allow us to nest JSON documents into database columns and you would translate the LINQ Expressions to the equivalent OData and/or SQL JSON query commands?

If you could reflect property names from dynamic types that would be nice, but alas ChatGPT tells me this is not possible ;)
https://chat.openai.com/share/78c21332-db16-4a97-a1ea-1d2ce786dc6b

But what if we just require concrete strong types for these nested queries but somehow designate a "sub-type" property so rather than using a generic Dictionary you would designate the MetaData with a strong type but with an Attribute that indicates that it's a nested JSON document?

**Reply by acidindustries** [reply]
It is an interesting idea indeed. I'm going to hack at it see if I can't come up with anything clever awaiting @adrianhall reply.

**Reply by adrianhall** [reply]
No approach for what I had in mind.  I need to take a look at the LINQ provider.

That being said, Azure Mobile Apps has - for quite a long time now - come to the end of its flexibility.  We know there are "problems" (generic) that are not easily achievable with a combination of OData and a custom client such as we have.  We are putting band-aids on problems at this point.

What we need is a new approach to doing offline synchronization.  I just haven't sat down to think solidly about this topic yet.  Offline synchronization is **hard**.  Sure, the simple case is simple indeed.  However, the generalized case, especially when you start considering foreign keys and referential integrity, is definitely hard.  If we were to consider how we make any object - no matter the shape - and any references - no matter the complexity - offline, and have the same query capabilities that LINQ provides, we will quickly come to the conclusion that (a) OData just isn't the right underlying protocol for this, and (b) None of the other standard protocols (GraphQL, gRPC) are suitable either.  We have to constrain the problem.  The constraints we chose for Azure Mobile Apps are showing their age and you are all coming up with the limitations of those constraints.

I don't have an easy fix for this.  Will do some thinking about it.

**Reply by amertlich** [reply]
There are similarities between what vlcn.io is doing [here](https://github.com/vlcn-io/cr-sqlite) and what Azure Mobile Apps is doing. At least as far as the data synchronization is concerned.  

VLCN's approach is using CRDTs as an extension to SQLite but is not fully baked yet. The advantage it has over AMA is that it's decentralized and supports multi-master replication. It also avoids manual merge conflicts by tracking changes at the column level and allowing users to specify the automated conflict resolution method (Last write wins, counter, fractional-index). 

Azure Mobile Apps is much more mature and includes features like partial sync (which is planed for cr-sqlite) and adds the LINQ ORM layer that can work both locally and on the remote database.

I'm not familiar with how VLCN would handle things like foreign keys and other referential integrity especially if they're enforced at the DB level. But I can say that we've been able to accommodate this even using the Azure Mobile Apps implementation as it sits currently, but with some complexity and abstractions that probably aren't suitable for most use-cases.

That being said, I think you may be being too harsh on AMA -- there's ways I could see it adopting some of the CRDT concepts and extending some of the LINQ query capabilities into nested JSON structures even if the ODATA protocol is only used as the descriptor language/syntax for remote queries -- although I could see GraphQL or something accommodating OpenAPI and/or JSONSchema being another approach.

**Reply by adrianhall** [reply]
I'm super critical of AMA, as you always should be of code you write.  It's the impetus to improve :-)

That being said, cr-sqlite is interesting.  I was also looking at [Fluid Framework](https://fluidframework.com/) which also uses CRDTs, and is a similar concept at a protocol level, and similarly enforces a specific database (their service, in the case of Fluid Framework).

So there is prior art, but - as I said at the beginning - it requires a complete re-architecture of the underlying transport.

**Comment by acidindustries** [comment]
I'm playing around with a custom ODataController to manage the sync. I can get the query working using the OData `[EnableQuery]` decorator on my QueryAsync method. Now I need to wrap PagedResults around it and integrate with the client but it seems to do the job for now. This is the actual "LINQ" query generated.

```csharp
System.Collections.Generic.List`1[testodata.Book].Where($it => (Convert(IIF((($it.Properties != null) AndAlso $it.Properties.ContainsKey("user")), $it.Properties.Item["user"], null), String) == value(Microsoft.AspNetCore.OData.Query.Container.LinqParameterContainer+TypedLinqParameterContainer`1[System.String]).TypedProperty))
```

Not sure if any of you has any input on using this solution. Seems like the latest OData implementation has most of the use cases figured out and could be a nice addition. I haven't looked at it extensively but my game plan is to have the EnableQuery filter running before the DataSyncController attribute so that the resulting filtered data gets wrapped into the proper wrappers. That might involve a rewrite of TableController to be implemented.

Actually, scratch that. Just adding `EnableQuery` on my TableController QueryAsync() seems to do the trick actually...
